### PR TITLE
Only use `git verify-tag` when verifying annotated tags

### PR DESCRIPTION
**Pull request by andreaso**
##### SUMMARY
The command `git verify-tag` only applies to annotated tags. When
verifying lightweight tags, which are more similar to non-moving
branches, one has to use `git verify-commit` instead.

##### ISSUE TYPE
 - Bugfix Pull Request

##### COMPONENT NAME
Git module

##### ANSIBLE VERSION
```
ansible 2.4.0 (devel 50f2313b47) last updated 2017/07/04 19:41:43 (GMT +200)
  config file = None
  configured module search path = [u'/home/andreas/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /home/andreas/workdir/ansible/lib/ansible
  executable location = /home/andreas/workdir/ansible/bin/ansible
  python version = 2.7.13 (default, Jan 19 2017, 14:48:08) [GCC 6.3.0 20170118]
```

##### ADDITIONAL INFORMATION
Let me demonstrate the difference when verifying annotated vs lightweight tags.

```
$ git tag
a_lightweight_tag
a_signed_annotated_tag
$
```
```
$ git verify-commit a_lightweight_tag
gpg: Signature made Tue Jul  4 22:12:04 2017 CEST
gpg:                using RSA key 210FE3E069ED67A8
gpg: Good signature from "Andreas Olsson <andreas@arrakis.se>" [full]
$ git verify-tag a_lightweight_tag
error: a_lightweight_tag: cannot verify a non-tag object of type commit.
$
```
```
$ git verify-commit a_signed_annotated_tag
error: a_signed_annotated_tag: cannot verify a non-commit object of type tag.
$ git verify-tag a_signed_annotated_tag
gpg: Signature made Tue Jul  4 22:14:29 2017 CEST
gpg:                using RSA key 210FE3E069ED67A8
gpg: Good signature from "Andreas Olsson <andreas@arrakis.se>" [full]
$
```

The `git for-each-ref ...` command I use behaves as follows.

```
$ git for-each-ref refs/tags/ --format '%(objecttype):%(refname:short)'
commit:a_lightweight_tag
tag:a_signed_annotated_tag
$
```

Using ':' as a separator is appropriate since that is one of the
characters not allowed in a Git reference name.

See also https://www.kernel.org/pub/software/scm/git/docs/git-check-ref-format.html

``git for-each-ref`` is a stable plumbing command. I have verified that it works consistently by testing it using the following Git versions.

* 1.9.1
* 2.7.4
* 2.13.0

### CODE REVIEW

**Comment by robinro** [comment]
Thanks @andreaso for this PR. The change looks good to me.

Can you add a test case for this? There is already a minimal test for tag verification in `test/integration/targets/git/tasks/tag-verification.yml`

**Comment by mattclay** [comment]
CI failure in integration tests: https://app.shippable.com/github/ansible/ansible/runs/28068/7/tests

```
{
    "changed": false, 
    "cmd": "gpg --keyserver keyserver.ubuntu.com --recv-key 0x210FE3E069ED67A8", 
    "failed": true, 
    "invocation": {
        "module_args": {
            "_raw_params": "gpg --keyserver keyserver.ubuntu.com --recv-key 0x210FE3E069ED67A8", 
            "_uses_shell": false, 
            "chdir": null, 
            "creates": null, 
            "executable": null, 
            "removes": null, 
            "warn": true
        }
    }, 
    "msg": "[Errno 2] No such file or directory", 
    "rc": 2
}
```

**Comment by andreaso** [comment]
Ok, testing added/improved!

@robinro

**Comment by robinro** [comment]
@andreaso Thanks for the update.

Do you think it is possible to do the test without an external repo?
So set up a local repo by running some shell commnds with a local key stored with the tests?
External dependencies have the tendency to break and are also bad for test performance.

**Comment by andreaso** [comment]
Good points! I'll look into it.

**Comment by andreaso** [comment]
Ok, here we go again! I have now rewritten the integration test to work on a local repo.

File: test/integration/targets/git/tasks/main.yml [hunk 1]
```diff
@@ -27,7 +27,11 @@
 - include: change-repo-url.yml
 - include: depth.yml
 - include: checkout-new-tag.yml
-- include: tag-verification.yml
+- include: gpg-verification.yml
+  when: >
+    not gpg_version.stderr and
+    gpg_version.stdout and
```
Comment (robinro) [inline test/integration/targets/git/tasks/main.yml:33]: There was a gpg_version >= 1.4.16 check in the old version. Is this not needed here?

**Review by robinro — APPROVED** [review]
LGTM
I assume the gpg version check is not needed (any more), since the CI passes on all supported distribution versions.

**Comment by robinro** [comment]
shipit

Thanks @andreaso for implementation.

File: test/integration/targets/git/tasks/main.yml (hunk 1 above)
Comment (andreaso) [reply test/integration/targets/git/tasks/main.yml:33]: That old check were kind of weird.

https://github.com/ansible/ansible/blob/v2.3.1.0-1/test/integration/targets/git/tasks/tag-verification.yml#L10-L11

The `or` logic resulting in a new enough GnuPG version negating the need for a new enough Git version.

**Comment by andreaso** [comment]
UPDATED: Please ignore this comment, and see the next one instead.

In case https://github.com/ansible/ansible/pull/26846 gets merged I would prefer not to have this pull request merged. Mucking around with someone's `~/.gnupg/` is not what I would categorize as non-destructive.

**Comment by andreaso** [comment]
Please ignore the previous comment. 

I have now improved the integration test to use a temporary `GNUPGHOME`, making it fully properly non-destructive. To clarify, I no longer see any conflict between this pull request and the https://github.com/ansible/ansible/pull/26846 PR.

I apologize for the noise.

**Pull request merged**